File: booking_service/src/core/usecase/get_user_usecase.go
package usecase

import (
	"booking_service/core/domain/constant"
	"booking_service/core/domain/entity"
	"booking_service/core/port"
	"booking_service/kernel/utils"
	"context"
	"encoding/json"
	"github.com/golibs-starter/golib/log"
)

type IGetUserUseCase interface {
	GetUserByID(ctx context.Context, ID int64) (*entity.UserEntity, error)
}

type GetUserUseCase struct {
	userPort  port.IUserPort
	cachePort port.ICachePort
}

func (g GetUserUseCase) GetUserByID(ctx context.Context, ID int64) (*entity.UserEntity, error) {
	// get from cache
	cachedUser, err := g.cachePort.GetFromCache(ctx, utils.BuildCacheKeyGetUser(ID))
	if err != nil && err.Error() != constant.ErrCacheKeyNil {
		log.Error(ctx, "Get User failed", err)
		return nil, err
	}
	if cachedUser != nil {
		response := entity.UserEntity{}
		err = json.Unmarshal([]byte(cachedUser.(string)), &response)
		if err != nil {
			log.Error(ctx, "Unmarshal error", err)
			return nil, err
		}
		return &response, nil
	}

	user, err := g.userPort.GetUserByID(ctx, ID)
	if err != nil {
		log.Error(ctx, "get user by id failed ", err)
		return nil, err
	}

	// set to cache
	go func() {
		err = g.cachePort.SetToCache(ctx, utils.BuildCacheKeyGetUser(ID), user, constant.DefaultCacheTTL)
		if err != nil {
			log.Error(ctx, "Set User to cache failed", err)
		}
	}()

	return user, nil
}

func NewGetUserUseCase(userPort port.IUserPort, cachePort port.ICachePort) IGetUserUseCase {
	return &GetUserUseCase{
		userPort:  userPort,
		cachePort: cachePort,
	}
}

File: authentication_service/src/main/java/huy/project/authentication_service/core/usecase/GetUserUseCase.java
package huy.project.authentication_service.core.usecase;

import huy.project.authentication_service.core.domain.constant.CacheConstant;
import huy.project.authentication_service.core.domain.constant.ErrorCode;
import huy.project.authentication_service.core.domain.entity.RoleEntity;
import huy.project.authentication_service.core.domain.entity.UserEntity;
import huy.project.authentication_service.core.domain.entity.UserRoleEntity;
import huy.project.authentication_service.core.exception.AppException;
import huy.project.authentication_service.core.port.ICachePort;
import huy.project.authentication_service.core.port.IUserPort;
import huy.project.authentication_service.core.port.IUserRolePort;
import huy.project.authentication_service.kernel.utils.CacheUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class GetUserUseCase {
    private final IUserPort userPort;
    private final IUserRolePort userRolePort;
    private final GetRoleUseCase getRoleUseCase;

    private final ICachePort cachePort;

    public UserEntity getUserById(Long userId) {
        // get from cache
        String key = CacheUtils.buildCacheKeyGetUserById(userId);
        UserEntity cachedUser = cachePort.getFromCache(key, UserEntity.class);
        if (cachedUser != null) {
            return cachedUser;
        }

        // get from db
        UserEntity user = userPort.getUserById(userId);
        if (user == null) {
            log.error("User not found");
            throw new AppException(ErrorCode.USER_NOT_FOUND);
        }

        setRolesToUser(user);

        // set to cache
        cachePort.setToCache(key, user, CacheConstant.DEFAULT_TTL);

        return user;
    }

    public UserEntity getUserByEmail(String email) {
        // get from cache
        String key = CacheUtils.buildCacheKeyGetUserByEmail(email);
        UserEntity cachedUser = cachePort.getFromCache(key, UserEntity.class);
        if (cachedUser != null) {
            return cachedUser;
        }

        // get from db
        UserEntity user = userPort.getUserByEmail(email);
        if (user == null) {
            log.error("user with email {} not found", email);
            throw new AppException(ErrorCode.USER_NOT_FOUND);
        }

        setRolesToUser(user);

        // set to cache
        cachePort.setToCache(key, user, CacheConstant.DEFAULT_TTL);

        return user;
    }

    private void setRolesToUser(UserEntity user) {
        List<UserRoleEntity> userRoles = userRolePort.getUserRolesByUserId(user.getId());
        List<Long> roleIds = userRoles.stream().map(UserRoleEntity::getRoleId).toList();
        List<RoleEntity> roles = getRoleUseCase.getAllRoles().stream()
                .filter(role -> roleIds.contains(role.getId()))
                .collect(Collectors.toList());
        user.setRoles(roles);
    }

    public long countAll() {
        return userPort.countAll();
    }
}
